wandb/integration/yolov8/utils.py

```diff
@@ -0,0 +1,173 @@
+"""
+Utilities for parsing and plotting Yolo results using wandb.
+"""
+import pathlib
+from copy import deepcopy
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import cv2
+import numpy as np
+import torch
+from ultralytics.yolo.utils.ops import xywh2xyxy
+
+import wandb
+from wandb import Image
+
+
+def convert_to_wb_images(
+    batch: Dict[str, Union[torch.Tensor, np.ndarray]],
+    cls: Union[torch.Tensor, np.ndarray],
+    bboxes: Union[torch.Tensor, np.ndarray],
+    masks: Optional[Union[torch.Tensor, np.ndarray]],
+    batch_idx: Union[torch.Tensor, np.ndarray],
+    names: Dict[int, str],
+) -> Tuple[List[Tuple[Image, Image]], List[Dict[Any, Union[float, Any]]]]:
+    """Utility to convert a batch of images, bounding boxes and masks from YOLO validator to list of wandb.Image objects
+    Additionally, returns a list of dictionaries with the average confidences for each class
+    """
+    images = batch["img"]
+    captions = batch["im_file"]
+    if isinstance(images, torch.Tensor):
+        images = images.cpu().float().numpy()
+    if isinstance(cls, torch.Tensor):
+        cls = cls.cpu().numpy()
+    if isinstance(bboxes, torch.Tensor):
+        bboxes = bboxes.cpu().numpy()
+    if isinstance(masks, torch.Tensor):
+        masks = masks.cpu().numpy().astype(int)
+    if isinstance(batch_idx, torch.Tensor):
+        batch_idx = batch_idx.cpu().numpy()
+
+    if np.max(images[0]) <= 1:
+        images *= 255  # de-normalise (optional)
+
+    out_images = []
+    class_set = wandb.Classes([{"name": v, "id": int(k)} for k, v in names.items()])
+
+    bs, _, h, w = images.shape
+    scores_data = []
+    for i in range(len(images)):
+        if len(cls) > 0:
```

Comment:
could we break this function into multiple helper functions it is quite hard to read it as is